### PR TITLE
Keep `attention_mask` as `None` in OPT's causal mask creation

### LINKED ISSUE #48924: OPT training with torch.compile decomposes SDPA into Triton kernels due to attention_mask
### Summary
When training OPT with torch.compile, an attention_mask=None input is  converted into an all-ones mask for positional embeddings. This makes the model pass an explicit mask to SDPA instead of using is_causal=True, preventing the optimized FlashAttention/oneDNN backward kernel and causing a 20–30% performance regression.

### Root Cause
In `src/transformers/models/opt/modeling_opt.py`:
```python
if attention_mask is None:
    attention_mask = torch.ones(
        (inputs_embeds.shape[0], all_seq_len), dtype=torch.bool, device=inputs_embeds.device
    )

causal_mask = create_causal_mask(
    config=self.config,
    input_tensor=inputs_embeds,
    attention_mask=attention_mask,  # <- passing non-None torch.ones mask
)
```

### PR DESCRIPTION
**Pull request by jiqing-feng**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=49002) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=49002)

Fixes #48924

Split out of #48975 on review request.

## Context

The same change was proposed in #48929 and closed, because it does not move the SDPA path on its own. That is correct, and this PR does not claim otherwise: the skip only takes effect together with the `masking_utils` fix in #48975. Measured on one A100 80GB with torch 2.14.0, `opt-125m`, batch 8, sequence 1024, bf16, sdpa, `torch.compile(dynamic=False)`, no padding mask:

| | fwd+bwd |
| --- | --- |
| `main` | 63.78 ms |
| this PR alone | 64.16 ms |
| with #48975 | 49.17 ms |

## Why it is still worth doing

Two things that may not be obvious from the earlier discussion.

The all-ones tensor is allocated on every forward, but the caller can already supply `position_ids`. At this call site `position_ids` is always computed before `self.embed_positions(...)`, so `OPTLearnedPositionalEmbedding.forward` never reads its `attention_mask` argument. Replacing it with zeros leaves the output bit-identical. The tensor exists only to be passed to something that ignores it, and to `create_causal_mask`, which is exactly where it does harm.

Separately, `create_causal_mask` is the one consumer that can tell the difference between "no mask" and "a mask that happens to be all ones". Handing it a synthetic mask discards information the caller gave us. Today only the compiled path observes this, but the constraint is not specific to `torch.compile`.

## Change

The dense tensor stays local, in `position_attention_mask`. The caller's `attention_mask` reaches `create_causal_mask` unchanged.

This is a no-op numerically. A dump of 30 intermediate tensors across fp32/bf16, unpadded, left and right padded inputs, in eager and under `torch.compile`, is bit-identical before and after.

## Tests

`test_attention_mask_is_not_overwritten_for_causal_mask` is new and fails on `main`.

`tests/models/opt`: 150 passed, 148 skipped.

## Note

Happy to close this if you would rather keep OPT as is until #48975 lands, or to fold it back in. It is also possible to drop the all-ones tensor entirely and derive the positions from `arange`, but that is a larger change and I left it out here.

### CODE REVIEW

**Comment by jiqing-feng** [comment]
Hi @vasqu . Would you please review this PR? Thanks!

File: tests/models/opt/test_modeling_opt.py [hunk 1]
```diff
@@ -261,6 +263,24 @@ def test_decoder_model_past_with_large_inputs(self):
         config_and_inputs = self.model_tester.prepare_config_and_inputs()
         self.model_tester.create_and_check_decoder_model_past_large_inputs(*config_and_inputs)
 
+    def test_attention_mask_is_not_overwritten_for_causal_mask(self):
```
Comment (vasqu) [inline tests/models/opt/test_modeling_opt.py]: let's make it mamba2 style like splitting into the tester but otherwise LGTM

File: tests/models/opt/test_modeling_opt.py [hunk 2]
```diff
@@ -261,6 +263,24 @@ def test_decoder_model_past_with_large_inputs(self):
...
+        model = OPTModel(config).to(torch_device).eval()
+
+        with patch(
+            "transformers.models.opt.modeling_opt.create_causal_mask", wraps=create_causal_mask
+        ) as mocked_create_causal_mask:
+            with torch.no_grad():
+                model(inputs_dict["input_ids"], attention_mask=None)
+
+        self.assertTrue(mocked_create_causal_mask.called)
+        self.assertIsNone(mocked_create_causal_mask.call_args.kwargs["attention_mask"])
```
Comment (vasqu) [inline tests/models/opt/test_modeling_opt.py]: Tbh wouldnt it also be possible to force the FA backend and run one forward? Just as alternative maybe, feels a bit more appropriate

**Review by vasqu — APPROVED** [review]
Just test design related

File: tests/models/opt/test_modeling_opt.py (hunk 1 above)
Comment (jiqing-feng) [reply tests/models/opt/test_modeling_opt.py]: Done. The body moved to `OPTModelTester.create_and_check_attention_mask_is_not_overwritten_for_causal_mask`.

File: tests/models/opt/test_modeling_opt.py (hunk 2 above)
Comment (jiqing-feng) [reply tests/models/opt/test_modeling_opt.py]: You are right, and it works now that #48975 is in `main`. Added as `test_compiled_forward_dispatches_to_flash_attention`: a compiled forward under `sdpa_kernel([SDPBackend.FLASH_ATTENTION])`, which only runs if the mask creation was skipped, since flash attention rejects any `attn_mask`.

| | eager | compiled |
| --- | --- | --- |
| main | passes | raises |
| this PR | passes | passes |

Eager cannot discriminate: `_ignore_causal_mask_sdpa` reads the all-ones tensor and returns `None` anyway. Only the compiled path sees the difference, because reading the values is declined while tracing.

I kept the mocked assertion as well, since the new test needs a GPU and is skipped on the CPU runners. Happy to drop it if you prefer just the one.

**Comment by vasqu** [comment]
Nice thanks merging
